File: src/ru/intellijeval/toolwindow/NewElementPopupAction.java
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package ru.intellijeval.toolwindow;

import com.intellij.icons.AllIcons;
import com.intellij.ide.IdeBundle;
import com.intellij.openapi.actionSystem.*;
import com.intellij.openapi.project.DumbAware;
import com.intellij.openapi.ui.popup.JBPopupFactory;
import com.intellij.openapi.ui.popup.ListPopup;
import org.jetbrains.annotations.NotNull;
import ru.intellijeval.Util;
import fork.com.intellij.openapi.fileChooser.actions.NewFolderAction;

/**
 * User: dima
 * Date: 13/08/2012
 */
class NewElementPopupAction extends AnAction implements DumbAware, PopupAction {
	public void actionPerformed(final AnActionEvent event) {
		showPopup(event.getDataContext());
	}

	protected void showPopup(DataContext context) {
		createPopup(context).showInBestPositionFor(context);
	}

	protected ListPopup createPopup(DataContext dataContext) {
		return JBPopupFactory.getInstance().createActionGroupPopup(
				IdeBundle.message("title.popup.new.element"),
				getGroup(),
				dataContext,
				false,
				true,
				false,
				null,
				-1,
				LangDataKeys.PRESELECT_NEW_ACTION_CONDITION.getData(dataContext)
		);
	}

	private ActionGroup getGroup() {
		return new ActionGroup() {
			@NotNull @Override
			public AnAction[] getChildren(AnActionEvent e) {
				return new AnAction[]{
						new NewFileAction("Groovy script", Util.GROOVY_FILE_TYPE_ICON),
						new NewFolderAction("Directory", "Directory", AllIcons.Nodes.Folder)
				};
			}
		};
	}

	public void update(AnActionEvent e) {
		Presentation presentation = e.getPresentation();
		presentation.setEnabled(true);
	}
}

File: src/main/liveplugin/implementation/actions/toolwindow/NewElementPopupAction.kt
package liveplugin.implementation.actions.toolwindow

import com.intellij.ide.IdeBundle
import com.intellij.openapi.actionSystem.*
import com.intellij.openapi.fileChooser.actions.NewFolderAction
import com.intellij.openapi.project.DumbAware
import com.intellij.openapi.ui.popup.JBPopupFactory
import com.intellij.openapi.ui.popup.ListPopup
import liveplugin.implementation.LivePluginPaths.groovyExamplesPath
import liveplugin.implementation.LivePluginPaths.kotlinExamplesPath
import liveplugin.implementation.actions.AddGroovyExamplesActionGroup
import liveplugin.implementation.actions.AddKotlinExamplesActionGroup
import liveplugin.implementation.actions.AddNewGroovyPluginAction
import liveplugin.implementation.actions.AddNewKotlinPluginAction
import liveplugin.implementation.actions.gist.AddPluginFromGistAction
import liveplugin.implementation.actions.git.AddPluginFromGitHubDelegateAction
import liveplugin.implementation.common.Icons.newFolderIcon
import liveplugin.implementation.common.IdeUtil.groovyFileType
import liveplugin.implementation.common.IdeUtil.kotlinFileType
import liveplugin.implementation.common.IdeUtil.textFileType
import liveplugin.implementation.common.IdeUtil.xmlFileType
import liveplugin.implementation.pluginrunner.groovy.GroovyPluginRunner.Companion.groovyScriptFile
import liveplugin.implementation.pluginrunner.groovy.GroovyPluginRunner.Companion.groovyTestScriptFile
import liveplugin.implementation.pluginrunner.kotlin.KotlinPluginRunner.Companion.kotlinScriptFile
import liveplugin.implementation.readSampleScriptFile

class NewElementPopupAction: AnAction(), DumbAware, PopupAction {
    override fun actionPerformed(event: AnActionEvent) {
        createPopup(event.dataContext).showInBestPositionFor(event.dataContext)
    }

    private fun createPopup(dataContext: DataContext): ListPopup =
        JBPopupFactory.getInstance().createActionGroupPopup(
            IdeBundle.message("title.popup.new.element"),
            livePluginNewElementPopup,
            dataContext, false, true, false, null, -1,
            LangDataKeys.PRESELECT_NEW_ACTION_CONDITION.getData(dataContext)
        )

    override fun update(event: AnActionEvent) {
        event.presentation.isEnabled = true
    }

    companion object {
        val livePluginNewElementPopup by lazy {
            DefaultActionGroup(
                { "New" },
                listOf(
                    NewKotlinFileAction(),
                    NewGroovyFileAction(),
                    NewTextFileAction(),
                    NewDirectoryAction(),
                    NewKotlinMainScript(),
                    NewGroovyMainScript(),
                    NewGroovyTestScript(),
                    NewPluginXmlScript(),
                    Separator.getInstance(),
                    AddNewKotlinPluginAction(),
                    AddNewGroovyPluginAction(),
                    AddPluginFromGistAction(),
                    AddPluginFromGitHubDelegateAction(),
                    AddKotlinExamplesActionGroup(),
                    AddGroovyExamplesActionGroup(),
                )
            ).also { it.isPopup = true }
        }
    }

    private class NewGroovyFileAction: NewFileAction("Groovy File", groovyFileType)

    private class NewKotlinFileAction: NewFileAction("Kotlin File", kotlinFileType)

    private class NewTextFileAction: NewFileAction("Text File", textFileType)

    private class NewDirectoryAction: NewFolderAction("Directory", "", newFolderIcon)

    private class NewGroovyMainScript: NewFileFromTemplateAction(
        text = groovyScriptFile,
        newFileName = groovyScriptFile,
        fileContent = readSampleScriptFile("$groovyExamplesPath/default-plugin.groovy"),
        fileType = groovyFileType,
    )

    private class NewKotlinMainScript: NewFileFromTemplateAction(
        text = kotlinScriptFile,
        newFileName = kotlinScriptFile,
        fileContent = readSampleScriptFile("$kotlinExamplesPath/default-plugin.kts"),
        fileType = kotlinFileType
    )

    private class NewGroovyTestScript: NewFileFromTemplateAction(
        text = groovyTestScriptFile,
        newFileName = groovyTestScriptFile,
        fileContent = readSampleScriptFile("$groovyExamplesPath/default-plugin-test.groovy"),
        fileType = groovyFileType
    )
}

class NewPluginXmlScript(
    fileContent: String = readSampleScriptFile("$kotlinExamplesPath/plugin.xml")
): NewFileFromTemplateAction(
    "plugin.xml",
    "plugin.xml",
    fileContent,
    xmlFileType
)
